test/lib/jdk/test/lib/serial/SerialObjectBuilder.java

```diff
@@ -0,0 +1,217 @@
+ * questions.
+ */
+
+package jdk.test.lib.serial;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.ObjectOutputStream;
+import java.io.UncheckedIOException;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import static java.io.ObjectStreamConstants.*;
+
+/**
+ * A basic builder of a serial object.
+ */
+public class SerialObjectBuilder {
+
+    private final ObjectOutputStream objectOutputStream;
+    private final ByteArrayOutputStream byteArrayOutputStream;
+
+    private record NameAndType<T>(String name, Class<T>type) { }
+
+    private String className;
+    private long suid;
+    private SerialObjectBuilder superClass;
+    private final LinkedHashMap<NameAndType<?>, Object> primFields = new LinkedHashMap<>();
+    private final LinkedHashMap<NameAndType<?>, Object> objectFields = new LinkedHashMap<>();
+
+    private SerialObjectBuilder() {
+        try {
+            byteArrayOutputStream = new ByteArrayOutputStream();
+            objectOutputStream = new ObjectOutputStream(byteArrayOutputStream);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    public static SerialObjectBuilder newBuilder(String className) {
+        return (new SerialObjectBuilder()).className(className);
+    }
+
+    private SerialObjectBuilder className(String className) {
+        this.className = className;
+        return this;
+    }
+
+    public SerialObjectBuilder suid(long suid) {
+        this.suid = suid;
+        return this;
+    }
+
+    public SerialObjectBuilder superClass(SerialObjectBuilder superClass) {
+        this.superClass = superClass;
+        return this;
+    }
+
+    public <T> SerialObjectBuilder addPrimitiveField(String name, Class<T> type, T value) {
+        if (!type.isPrimitive())
+            throw new IllegalArgumentException("Unexpected non-primitive field: " + type);
+        primFields.put(new NameAndType<>(name, type), value);
+        return this;
+    }
+
+    public <T> SerialObjectBuilder addField(String name, Class<T> type, T value) {
+        if (type.isPrimitive())
+            throw new IllegalArgumentException("Unexpected primitive field: " + type);
+        objectFields.put(new NameAndType<>(name, type), value);
+        return this;
+    }
+
+    private static int getPrimitiveSignature(Class<?> cl) {
+        if (cl == Integer.TYPE) return 'I';
+        else if (cl == Byte.TYPE) return 'B';
+        else if (cl == Long.TYPE) return 'J';
+        else if (cl == Float.TYPE) return 'F';
+        else if (cl == Double.TYPE) return 'D';
+        else if (cl == Short.TYPE) return 'S';
+        else if (cl == Character.TYPE) return 'C';
+        else if (cl == Boolean.TYPE) return 'Z';
+        else throw new InternalError();
+    }
+
+    private static void writeUTF(DataOutputStream out, String str) throws IOException {
+        int utflen = str.length(); // assume ASCII
```

Comment:
Add an assert to confirm its ASCII, or update the code to correctly estimate the number of bytes.
Possibly by using DOS to write to a temporary BAOS.